### PR TITLE
Use jitstd::list in the LSRA.

### PR DESCRIPTION
**Pull request by pgavlin**
This removes the only usage of ArrayList in the JIT.

### CODE REVIEW

**Comment by pgavlin** [comment]
@dotnet/jit-contrib @CarolEidt PTAL

File: src/jit/jitstd/list.h [hunk 1]
```diff
@@ -68,6 +68,7 @@ class list
         bool operator!=(const const_iterator& it) const;
         const T& operator*() const;
         const T* operator&() const;
+        const T* operator->() const;
```
Comment (CarolEidt) [inline src/jit/jitstd/list.h]: Can you explain how this is being used? It's not clear to me.

**Comment by CarolEidt** [comment]
My concern about these changes is that there is now an inconsistency wrt to the usage of pointers vs. references for Intervals and RefPositions. The former convention was that anything with an "interval" name was a pointer to an interval, and anything with a "refposition" or "rp" name was a pointer to a RefPosition. Now it is inconsistent - perhaps we should change over to using all references, but I'm not happy with the inconsistency (and, in general, pointers are used all over the JIT, so perhaps this is a conventions kind of question).

File: src/jit/jitstd/list.h (hunk 1 above)
Comment (pgavlin) [inline src/jit/jitstd/list.h]: LSRA currently uses the following patter in a few places:

```
ItemIterator iter = someList.begin();

item* i;
for (i = *iter, ++iter; i != nullptr && !i->someCondition(); i = *iter, ++iter)
{
   i->someMethod();
}
```

These uses have been changed to the following pattern:

```
jiststd::list<item>::iterator i = someList.begin()
for (; i != someList.end() && !i->someCondition(); ++i)
{
   i->someMethod(); // This statement uses jitstd::list<item>::operator-> to access the list entry at the current position
}
```

**Comment by pgavlin** [comment]
> Now it is inconsistent - perhaps we should change over to using all references, but I'm not happy with the inconsistency (and, in general, pointers are used all over the JIT, so perhaps this is a conventions kind of question).

The use of references is somewhat necessary with the change to jitstd::list<T>, since its elements are accessed as `T&` during iteration (in contrast to `StructArrayList<T>`, whose iterator returns values of type `T*`). If you're very concerned with the difference, I can change the few sites where wide-scale changes were necessary to something like the following:

```
for (auto& refPosition : refPositions)
{
    RefPosition* currentRefPosition = &refPosition;
    ...
}
```

which also has the side-effect of producing a smaller diff in some places.

**Comment by CarolEidt** [comment]
@pgavlin I think it would probably be a good idea to localize the changes as you suggest. It's still a little weird and inconsistent, but preserves a consistent usage for the bulk of the cases.

**Comment by pgavlin** [comment]
Done. There are a couple uses of references for very short loops (mainly used as part of LSRA dumps)  that I left in: in these cases I felt that localizing the changes resulted in less clear code.

**Comment by CarolEidt** [comment]
Looks great - thanks. A big improvement over the old macros :+1:

**Comment by pgavlin** [comment]
@mmitche it looks like there is some problem with package restore on the x64 Debug build.

**Comment by AndyAyersMS** [comment]
Interesting, I hit that same error on my local laptop. 

It looks like when package restores fail, we just carry on, and end up with gazillions of compile errors as a result. Interactively this wouldn't be so bad, but msbuild is almost impossible to killl from the command line, so you end up having to hunt it down the task manager.

**Comment by pgavlin** [comment]
Given that this change has passed a DDR and local GH testing, I'm going to go ahead and merge.

**Comment by mmitche** [comment]
@pgavlin @weshaggard @AndyAyersMS @ericstj Any idea why this happens?  We shouldn't carry on after restore failure.

**Comment by AndyAyersMS** [comment]
Building tests, I see entries like 
```
         [12:32:01.80] Restoring all packages...
         (set CORE_ROOT=) & "C:\repos\coreclr\tests\..\Tools\dotnetcli/bin/dotnet.exe" restore --packages "C:\repos\coreclr\tests\..\packages" --source https://dotnet.myget.org/F/dotnet-core/api/v3/index.json --source https://www.myget.org/F/nugetbuild/api/v3/index.json --source https://api.nuget.org/v3/index.json "C:\repos\coreclr\tests\src"
     1>EXEC : warning : Failed to retrieve information from remote source 'https://myget-2e16.kxcdn.com/artifacts/nugetbuild/nuget/v3/flatcontainer/system.threading.tasks/index.json'. [C:\repos\coreclr\tests\build.proj]
         The previous error was converted to a warning because the task was called with ContinueOnError=true.
     1>EXEC : warning : An error occurred while sending the request. [C:\repos\coreclr\tests\build.proj]
```
and then we keep going....

**Pull request merged**